modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpServerTransportTests.java

```diff
@@ -644,11 +649,214 @@ public void dispatchBadRequest(final RestChannel channel, final ThreadContext th
+                networkService,
+                threadPool,
+                xContentRegistry(),
+                dispatcher,
+                clusterSettings,
+                new SharedGroupFactory(settings),
+                Tracer.NOOP,
+                TLSConfig.noTLS(),
+                null,
+                successHeadersValidator
+            ) {
+                @Override
+                protected void populatePerRequestThreadContext(RestRequest restRequest, ThreadContext threadContext) {
+                    assertThat(threadPool.getThreadContext().getHeader(contextHeaderReference.get()), nullValue());
+                    assertThat(threadPool.getThreadContext().getTransient(contextHeaderReference.get()), nullValue());
+                    ThreadContext.StoredContext storedAuthenticatedContext = HttpHeadersAuthenticatorUtils.extractAuthenticationContext(
+                        restRequest.getHttpRequest()
+                    );
+                    assertThat(storedAuthenticatedContext, notNullValue());
+                    // restore validation context
+                    storedAuthenticatedContext.restore();
+                    assertThat(
+                        threadPool.getThreadContext().getHeader(contextHeaderReference.get()),
+                        is(contextHeaderValueReference.get())
+                    );
+                    assertThat(
+                        threadPool.getThreadContext().getTransient(contextHeaderReference.get()),
+                        is(contextHeaderValueReference.get())
+                    );
+                }
+            }
+        ) {
+            transport.start();
+            final TransportAddress remoteAddress = randomFrom(transport.boundAddress().boundAddresses());
+            for (HttpMethod httpMethod : List.of(HttpMethod.GET, HttpMethod.POST, HttpMethod.PUT, HttpMethod.DELETE, HttpMethod.PATCH)) {
+                httpMethodReference.set(httpMethod);
+                urlReference.set(
+                    "/"
+                        + randomAlphaOfLengthBetween(4, 8)
+                        + "?X-"
+                        + randomAlphaOfLengthBetween(4, 8)
+                        + "="
+                        + randomAlphaOfLengthBetween(4, 8)
+                );
+                requestHeaderReference.set("X-" + randomAlphaOfLengthBetween(4, 8));
+                requestHeaderValueReference.set(randomAlphaOfLengthBetween(4, 8));
+                try (Netty4HttpClient client = new Netty4HttpClient()) {
+                    FullHttpRequest request = new DefaultFullHttpRequest(
+                        HttpVersion.HTTP_1_1,
+                        httpMethodReference.get(),
+                        urlReference.get()
+                    );
+                    request.headers().set(requestHeaderReference.get(), requestHeaderValueReference.get());
+                    FullHttpResponse response = client.send(remoteAddress.address(), request);
```

Comment:
(only commenting here, not suggesting to do this here).... it might be overkill, but it would be nice to have a multi-threaded client sending a bunch of requests with some expected to pass validation and others expected to fail all attempting to send requests concurrently. Assuming the test isn't too articical it could help catch any possible  bugs if the thread context ever bleeds between requests.

Reply:
The validation is all on the sole thread per channel, so a multithreaded client is uninteresting.
Also, all validation is entierely segragated by channels so a multi channel test is also uninteresting.
I'll see to add tests where subsequent requests on the same channel use different credentials, some of which fail validation.

Reply:
> I'll see to add tests where subsequent requests on the same channel use different credentials, some of which fail validation.

That is mostly what I was suggesting, but the multithreaded part was to attempt to have concurrent requests with different credentials in the same channel at the same time. Thinking abit more, i think the only way to do that would be to have ensure you are (http 1.1) pipeline requests and/or a chunked request (i.e. request broken up with 100 continue statuses).

Reply:
I've pushed [815e523](https://github.com/elastic/elasticsearch/pull/95112/commits/815e5234b3445939f319733df7cfdae919575b8b) that tests multiple requests with validated differently on the same channel.